fairroot/base/steer/FairRunAna.cxx

```diff
@@ -432,7 +433,7 @@ void FairRunAna::RunEventReco(Int_t Ev_start, Int_t Ev_end)
         fRootManager->StoreWriteoutBufferData(fRootManager->GetEventTime());
         fTask->ExecuteTask("");
 
-        FillEventHeader();
+        FillEventHeader(GetEventHeader());
```

Comment:
Hi. By doing so, my idea was that the function FillEventHeader takes the event header as a parameter and therefore is independent of the event header intrinsically. It's a very subtle change. IMO, it's more readable and less error-prone. Otherwise, one possibility to make an error is to change the event header inside the `FillEventHeader()`.

But now I think we should get just rid of it. Instead we could add a getter for the fSource and directly fill the event header from the source. I think it's more clearer in this way.

Reply:
I learned about the Facade pattern from @dennisklein. And yes, I think, `FillEventHeader()` and `GetEvtHeaderRunId()` are facades.

And we're not trying to hide anything away. First, these are internal member functions. Not meant for public use (at least not yet). Second, the full source is available anyway.

I tried to improve documentation with #1413.

Reply:
From my understanding, the Facade design pattern involves with a Facade object that provides a simpler public interface to a complicated underlying system. But here the function `FillEventHeader()` just provide an additional way to modify its private member variable.

Protected functions are still not private, which means any derived class could expose the functionality to the public.

However the question remains: there are two paths/ways to do one thing. This could further cause additional problems:

1. Confusing to users. 
   Which way should I choose? 
  ```cpp
  auto* evtHeader = run->GetEventHeader();
  auto* source = run->GetSource();
  source->FillEventHeader(evtHeader);
  ```
  or
  ```cpp
  run->FillEventHeader();
  ```
2. Bug-prone and difficult to maintain. 
    Supposing that in the future there is a change for the function `FillEventHeader()`, the code using the second way could have this new implementation. But other code using the first method still has the old implementation even though these two method are meant to do exactly the same thing.

Reply:
> Protected functions are still not private, which means any derived class could expose the functionality to the public.

Those functions have been declared (via documentation) as internal in #1413. If someone outside FairRoot uses them, they're entirely on their own. If this is actually a concern, we could mark them as `private:` and use `friend` to allow all the relevant users inside FairRoot. But TBH I don't like that either: At some point, I like it, if things are private even inside FairRoot.

Point (1) then isn't relevant: Only possibly confusing to FairRoot developers. My take on that one: Use the shorter API, so that the caller is easier to read (which is really an important point in our current codebase).

Point (2) is also only relevant to FairRoot developers. But that conern is a problematic point of the facade pattern at all. So yes, FairRoot devs have to take care.